analysis/import/SCMLogParser/src/main/kotlin/de/maibornwolff/codecharta/importer/scmlogparser/parser/CommitCollector.kt

```diff
@@ -10,70 +11,83 @@ import java.util.stream.Collector
-        filename: String
-    ): VersionControlledFile? {
-        return versionControlledFiles.firstOrNull { it.filename == filename }
-    }
+            val renameName = renamesMap[possibleConflictName]
+            val VCFName = if (renameName == null) possibleConflictName else renameName
 
-    private fun addYetUnknownFile(
-        versionControlledFiles: MutableList<VersionControlledFile>,
-        filenameOfYetUnversionedFile: String
-    ): Boolean {
-        val missingVersionControlledFile = VersionControlledFile(filenameOfYetUnversionedFile, metricsFactory)
-        return versionControlledFiles.add(missingVersionControlledFile)
-    }
+            when (it.type) {
 
-    private fun addCommitMetadataToMatchingVersionControlledFiles(
-        commit: Commit,
-        versionControlledFiles: List<VersionControlledFile>
-    ) {
-        commit.filenames.mapNotNull { findVersionControlledFileByFilename(versionControlledFiles, it) }
-                .forEach { it.registerCommit(commit) }
+                Modification.Type.ADD -> {
+                    val file = versionControlledFiles[possibleConflictName]
+                    if (file == null) {
+                        val missingVersionControlledFile = VersionControlledFile(possibleConflictName, metricsFactory)
+                        versionControlledFiles[possibleConflictName] = missingVersionControlledFile
+                        missingVersionControlledFile.registerCommit(commit, it)
+                    } else {
+                        versionControlledFiles[VCFName]!!.registerCommit(commit, it)
+                    }
+                }
+
+                Modification.Type.DELETE -> {
+                    val filename = renamesMap[possibleConflictName]
+                    versionControlledFiles.remove(if (filename == null) possibleConflictName else filename)
+                    renamesMap.remove(possibleConflictName)
+                }
+                Modification.Type.RENAME -> {
+                    var newVCFFileName = it.currentFilename
+                    if (versionControlledFiles.containsKey(it.currentFilename)) {
+                        val marker = nameConflictsMap[it.currentFilename]
+                        val newMarker = if (marker != null) marker + 1 else 0
+                        newVCFFileName = it.currentFilename + "_\\0_" + newMarker
+                        nameConflictsMap[it.currentFilename] = newMarker
+                    }
+                    if (renameName != null) {
+                        renamesMap.remove(possibleConflictName)
+                        renamesMap[newVCFFileName] = renameName
+                    } else {
+                        renamesMap[newVCFFileName] = it.oldFilename
+                    }
+                    versionControlledFiles[VCFName]?.registerCommit(commit, it)
+                }
+                else -> versionControlledFiles[VCFName]?.registerCommit(commit, it)
+            }
+        }
```

Comment:
This logic is not trivial and I suggest to outline the logic as big comment above the code. That way it's easier for others to grasp how it works and it's also easier to identify logic bugs.

Reply:
Ok I will add a comment with example, maybe writing it will help detecting the logical error as well.